xivo_confgen/plugins/pjsip_conf.py

```diff
@@ -588,7 +601,20 @@ def _get_transport_wss(self):
 
         return self._get_base_udp_transport('wss')
 
-    def _add_from_mapping(self, fields, mapping, config):
+    @staticmethod
+    def _add_pjsip_options(fields, options, config, from_list=False):
```

Comment:
This function should check of the field presence before adding it. It creates duplicate entries in the sections of the config file, since for an endpoint the following fields are already added:
(pjsip_conf.py lines 289-305)
```python
fields = [
            ('type', 'endpoint'),
            ('context', user_dict['context']),
            ('aors', user_sip[0].name),
            ('auth', user_sip[0].name),
            ('set_var', 'XIVO_ORIGINAL_CALLER_ID={callerid}'.format(**user_dict)),
            ('set_var', 'TRANSFER_CONTEXT={}'.format(user_sip.context)),
            ('set_var', 'PICKUPMARK={}%{}'.format(user_sip.number, user_sip.context)),
            ('set_var', 'XIVO_USERID={}'.format(user_sip.user_id)),
            ('set_var', 'XIVO_USERUUID={}'.format(user_sip.uuid)),
            ('set_var', 'WAZO_TENANT_UUID={}'.format(user_sip.UserSIP.tenant_uuid)),
            ('set_var', 'WAZO_CHANNEL_DIRECTION=from-wazo'),
        ]

        for key, value in all_options:
            if key in ('allow', 'disallow'):
                self._add_option(fields, (key, value))
```

Reply:
It is ok to have some options multiple times in a given section. I added the pjsip fields after the conversion to be able to override the result of the sip to pjsip conversion.
e.g. if you use nat=comedia, rtp_symmetric=yes, rewrite_contacts=yes the generated file will be

rtp_symmetric=yes  ; from the nat option
rtp_symmetric=yes  ; from the hard coded rtp_symmetric
rewrite_contacts=yes  ; from the hard coded rewrite_contacts

That way the auto-generated fields are ignored when you know which fields you need in the pjsip config.

some other fields are also "accumulating" like `allow` and `set_var`

Did you have a specific field that caused a problem?

Reply:
No, I did not have problems. I only saw those duplicates when doing a diff between a config file before and a config file after the changes and I thought that it was not a wanted behaviour.